Review

rmnkubyshkin opened the pull request:
pr from branch review to main

rmnkubyshkin committed:
feat: Realized routers, store, slices, reducers.

	1. Rezlized routers, like Ingredients, DetailsIngredients,
	Feed, FeedInfo, OrderInfo
	2. Realized sending parameters throught this components
	3. Rezlized store, ingredientSlise, feedSlise, reducers and
	   selectors for that slices

rmnkubyshkin committed:
q: not working constructorSlice

rmnkubyshkin committed:
q: not updated order state

rmnkubyshkin committed:
feat: routing, api for auth,order,constructor components

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,170 @@
+        <Route path='*' element={<NotFound404 />} />
+        <Route path='/' element={<ConstructorPage />} />
+        <Route
+          path='/login'
```
Надо исправить: не работает счетчик ингредиентов

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,170 @@
+  onlyUnAuth,
+  children
+}: ProtectedRouteProps) => {
+  const dispatch: AppDispatch = useDispatch();
```
Можно лучше: разные запросы стоит выносить по разным папкам

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,170 @@
+  );
+  const user = useSelector((state: RootState) => state.user.user);
+  const _getUser = async () => await dispatch(checkUserAuth()).unwrap();
+  _getUser();
```
Надо исправить: `dispatch` это эффект, так что должен обязательно вызываться в `useEffect()`

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,170 @@
+            <Modal
+              title='IngredientDetails'
+              onClose={() => {
+                handleCloseIngredient();
```
Надо исправить: модальные окна должны оставаться модальными окнами. То есть должна быть видна подложка с текущим состоянием страницы

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,170 @@
+          element={
+            <Modal
+              title='OrderInfo'
+              onClose={() => {
```
Надо исправить: вместо `OrderInfo` должен быть номер заказа

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,170 @@
+  const dispatch: AppDispatch = useDispatch();
+  const location = useLocation();
+  const isAuthChecked = useSelector(
+    (state: RootState) => state.user.isAuthChecked
```
Можно лучше: стоит типизировать хуки, чтобы каждый раз не писать тип 

https://react-redux.js.org/using-react-redux/usage-with-typescript#define-typed-hooks

kyzinatra commented on src/pages/login/login.tsx:
```diff
@@ -1,12 +1,22 @@
 export const Login: FC = () => {
+  const dispatch: AppDispatch = useDispatch();
+  const navigate = useNavigate();
   const [email, setEmail] = useState('');
```
Можно лучше: стоит удалить не используемые хуки

kyzinatra commented on src/pages/login/login.tsx:
```diff
@@ -1,12 +1,22 @@
+  const dispatch: AppDispatch = useDispatch();
+  const navigate = useNavigate();
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
```
Можно лучше: Чтобы не повторять логику для подвязки инпутов можно создать хук `useForm()`, который бы позволил делать это более коротко. Пример:

```js

const [values, onChange] = useForm({email: "", password: ""})

values // {email: string, password: string}
onChange // (event) => void

```

Примерно вот так можнт выглядеть реализация хука. Тут конечно можно дальше добавлять обработку ошибок, загрузки итп итд

```js

function useForm<T extends { [key: string]: string }>(baseForm: T) {
  const [form, setForm] = useState<T>(baseForm);

  function handleChange(e: ChangeEvent<HTMLInputElement>) {
    const element = e.target;
    setForm((pastForm) => ({ ...pastForm, [element.name]: element.value }));
  }

  return [form, handleChange];
}


```

kyzinatra commented:
Выглядит очень круто! Видно, что проделана большая работа. Но все же есть моменты на которые стоит обратить внимание.
Подробнее: 

**Что важно исправить:**

-  не работает счетчик ингредиентов
-  `dispatch` это эффект, так что должен обязательно вызываться в `useEffect()`
-  модальные окна должны оставаться модальными окнами. То есть должна быть видна подложка с текущим состоянием страницы
-  вместо `OrderInfo` должен быть номер заказа


**Можно лучше:**

-  разные запросы стоит выносить по разным папкам
-  стоит типизировать хуки, чтобы каждый раз не писать тип 

https://react-redux.js.org/using-react-redux/usage-with-typescript#define-typed-hooks
-  стоит удалить не используемые хуки
-  Чтобы не повторять логику для подвязки инпутов можно создать хук `useForm()`, который бы позволил делать это более коротко. Пример:

```js

const [values, onChange] = useForm({email: "", password: ""})

values // {email: string, password: string}
onChange // (event) => void

```

Примерно вот так можнт выглядеть реализация хука. Тут конечно можно дальше добавлять обработку ошибок, загрузки итп итд

```js

function useForm<T extends { [key: string]: string }>(baseForm: T) {
  const [form, setForm] = useState<T>(baseForm);

  function handleChange(e: ChangeEvent<HTMLInputElement>) {
    const element = e.target;
    setForm((pastForm) => ({ ...pastForm, [element.name]: element.value }));
  }

  return [form, handleChange];
}


```


**Что понравилось:**

- Хорошая работа с формами

**Обратите внимание, что работа принимается только после исправления всех `«Надо исправить»`.**

rmnkubyshkin committed:
fix: Bugs by iteration

	1. Typed useDispatch hook
	2. Moved protected-route components to individual folder and files .tsx
	3. Fix counter by Ingredient component, when it added to
	   order
	4. Calling useDispatch moved to useEffect
	5. Fix viewing Modal component, at the moment we can see
	   background page, it works as Modal component
	6. Added viewing number of Order in Modal View, on the top
	7. Deleted unused hooks

kyzinatra commented:
Вы отлично потрудились.
Поздравляю! Ваша работа принята.
Удачного дальнейшего обучения.

kyzinatra commented on src/components/burger-ingredient/burger-ingredient.tsx:
```diff
@@ -1,21 +1,17 @@
-    const dispatch: AppDispatch = useDispatch();
+    const dispatch = useAppDispatch();
+    useEffect(() => {
+      count += 1;
```
Кажется этот эффект тут не нужен. Реакт запрещает мутировать пропсы
